tests/unit/backend/workflow/test_validation.py

```diff
@@ -0,0 +1,111 @@
+import json
+from json import JSONDecodeError
+from jsonschema.exceptions import ValidationError
+import pytest
+from pathlib import Path
+from imars3d.backend.workflow.validate import JSONValid, JSONValidationError, SCHEMA
+
+JSON_DATA_DIR = Path(__file__).parent.parent.parent.parent / "data" / "json"
+GOOD_FILE = JSON_DATA_DIR / "good.json"
+ILL_FORMED_FILE = JSON_DATA_DIR / "ill_formed.json"
+
+
+class MockContainer:
+    """Mock class for mimicing how json validation will work in practice"""
+
+    config = JSONValid(schema=SCHEMA)
+
+    def __init__(self, obj):
+        self.config = obj
+
+
+def load_file(filepath: Path) -> str:
+    """Utility function for loading json from file"""
+
+    with open(filepath, "r") as handle:
+        data = handle.read()
+    return data
+
+
+# tests from file
+
+
+def test_file_not_exist():
+    with pytest.raises(FileNotFoundError):
+        MockContainer(Path("non-existant-file"))
+
+
+def test_file_ill_formed():
+    print("Testing file", ILL_FORMED_FILE)
+    with pytest.raises(JSONDecodeError):
+        MockContainer(ILL_FORMED_FILE)
+
+
+def test_file_good():
+    print("Testing file", GOOD_FILE)
+    MockContainer(GOOD_FILE)
+
+
+# tests from string
+
+
+def test_string_empty():
+    with pytest.raises(JSONDecodeError):
+        MockContainer("")
+
+
+def test_string_ill_formed():
+    doc = load_file(ILL_FORMED_FILE)
+    with pytest.raises(JSONDecodeError):
+        MockContainer(doc)
+
+
+def test_string_good():
+    doc = load_file(GOOD_FILE)
+    obj = MockContainer(doc)
+    # verify that the instrument is as expected
+    assert obj._json["instrument"] == "CG1D"
+
+
+def test_string_missing_tasks():
+    doc = load_file(GOOD_FILE)
+    json_obj = json.loads(doc)
+    del json_obj["tasks"]
+    with pytest.raises(JSONValidationError):
+        MockContainer(json.dumps(json_obj))
+
+
+@pytest.mark.parametrize(
+    "facility,instrument", [("hfir", "CG1D"), ("hfir", "CG1D"), ("HFIR", "SNAP"), ("HFIR", "junk"), ("SNS", "junk")]
```

Comment:
why are we testing `("hfir", "CG1D")` twice here?

Reply:
I think I'll change the second tuple to `("HFIR", "cg1d")`. According to `_validate_facility_and_instrument`, both the facility and the instrument must be uppercase.